File: src/wy_qcos/transpiler/cmss/circuit/dag_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# ----------------------------------------------------------------------
# Copyright© 2024-2025 China Mobile (SuZhou) Software Technology Co.,Ltd.
#
# qcos is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions
# of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#         http://license.coscl.org.cn/MulanPSL2
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS,
#     WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.
# ----------------------------------------------------------------------

from collections.abc import Iterable


class DAGNode:
    """Parent class for DAGOpNode, DAGInNode, and DAGOutNode."""

    __slots__ = ["_node_id"]

    def __init__(self, nid=-1):
        """Create a node."""
        self._node_id = nid

    def __lt__(self, other):
        return self._node_id < other._node_id

    def __gt__(self, other):
        return self._node_id > other._node_id

    def __str__(self):
        return str(id(self))


class DAGOpNode(DAGNode):
    """Object to represent an Gate at a node in the DAGCircuit."""

    __slots__ = ["op", "qargs", "cargs", "sort_key"]

    def __init__(
        self, op, qargs: Iterable[int] = (), cargs: Iterable[int] = ()
    ):
        """Create an Gate node."""
        super().__init__()
        self.op = op
        self.qargs = tuple(qargs)
        self.cargs = tuple(cargs)
        self.sort_key = str(self.qargs)

    @property
    def name(self):
        """Returns the Gate name corresponding to the op for this node."""
        return self.op.name

    @name.setter
    def name(self, new_name):
        """Sets the Gate name corresponding to the op for this node."""
        self.op.name = new_name

    def __repr__(self):
        """Returns a representation of the DAGOpNode."""
        return (
            f"DAGOpNode(op={self.op}, qargs={self.qargs}, cargs={self.cargs})"
        )


class DAGInNode(DAGNode):
    """Object to represent an incoming wire node in the DAGCircuit."""

    __slots__ = ["wire", "sort_key"]

    def __init__(self, wire):
        """Create an incoming node."""
        super().__init__()
        self.wire = wire
        self.sort_key = str([])

    def __repr__(self):
        """Returns a representation of the DAGInNode."""
        return f"DAGInNode(wire={self.wire})"


class DAGOutNode(DAGNode):
    """Object to represent an outgoing wire node in the DAGCircuit."""

    __slots__ = ["wire", "sort_key"]

    def __init__(self, wire):
        """Create an outgoing node."""
        super().__init__()
        self.wire = wire
        self.sort_key = str([])

    def __repr__(self):
        """Returns a representation of the DAGOutNode."""
        return f"DAGOutNode(wire={self.wire})"

File: src_cpp/transpiler_cpp/src/circuit/dag_node.cpp
/*
 * ----------------------------------------------------------------------
 * Copyright© 2024-2026 China Mobile (SuZhou) Software Technology Co.,Ltd.
 *
 * qcos is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions
 * of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS,
 *      WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 * ----------------------------------------------------------------------
 */

#include "circuit/dag_node.h"

#include <sstream>

namespace qcos {

std::string vector_to_string(const std::vector<int>& values) {
  std::ostringstream oss;
  oss << "[";
  for (size_t i = 0; i < values.size(); ++i) {
    if (i > 0) {
      oss << ", ";
    }
    oss << values[i];
  }
  oss << "]";
  return oss.str();
}

DAGOpNode::DAGOpNode(std::shared_ptr<BaseOperation> op_,
                     std::vector<int> qargs_, std::vector<int> cargs_)
    : op(std::move(op_)), qargs(std::move(qargs_)), cargs(std::move(cargs_)) {
  sort_key_ = vector_to_string(qargs);
}

std::string DAGOpNode::repr() const {
  std::ostringstream oss;
  oss << "DAGOpNode(op=" << op->name << ", qargs=" << vector_to_string(qargs)
      << ", cargs=" << vector_to_string(cargs) << ")";
  return oss.str();
}

}  // namespace qcos
